<!DOCTYPE html>
<html>
<head>
    <!-- Standard Meta -->
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0">

    <title>Genome-wide STR Truth Set</title>
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/semantic-ui/2.4.1/semantic.min.css" integrity="sha512-8bHTC73gkZ7rZ7vpqUQThUDhqcNFyYi2xgDgPDHc+GXVGHXq+xPjynxIopALmOPqzo9JZj0k6OqqewdGO3EsrQ==" crossorigin="anonymous" referrerpolicy="no-referrer" />

    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=G-G7WEXHL9KW"></script>
    <script>
        window.dataLayer = window.dataLayer || [];
        function gtag(){dataLayer.push(arguments);}
        gtag('js', new Date());

        gtag('config', 'G-G7WEXHL9KW');
    </script>
    <style>
        /* tool comparison plot viewer */
        .ui.form {
            font-size: 16px;
        }
        .ui.form label {
            font-size: 16px;
        }
    </style>
</head>
<body>
<div class="ui stackable grid" style="padding: 0px 0px 0px 30px">
    <div class="ui row">
        <div class="ui eight wide column" style="min-width:820px">
            <div class="ui form" style="padding: 30px 30px 0px 30px">
                <div class="inline fields">
                    <label style="min-width:160px">Plot Type:</label>
                    <div class="field"><div class="ui radio checkbox"><input type="radio" value="AccuracyVsQ" name="plot-type" checked="checked"><label><b>Accuracy vs. Q Threshold</b></label></div></div>
                    <div class="field"><div class="ui radio checkbox"><input type="radio" value="AccuracyByAlleleSize" name="plot-type"><label><b>Accuracy by Allele Size</b></b></label></div></div>
                </div>
            </div>
            <ul>
                <li>
                    <div class="ui form">
                        <div class="inline fields">
                            <label style="min-width:150px">Motif Size:</label>
                            <div style="min-width: 70px">[<a class="prev-button" id="prev-motif-size" href="#">←</a> <a class="next-button" id="next-motif-size" href="#">→</a>]</div>
                            <div class="field"><div class="ui radio checkbox"><input type="radio" value="all_motifs" name="motif-size"><label>all</label></div></div>
                            <div class="field"><div class="ui radio checkbox"><input type="radio" value="2to6bp_motifs" name="motif-size" checked="checked"><label>2-6bp</label></div></div>
                            <div class="field"><div class="ui radio checkbox"><input type="radio" value="7to24bp_motifs" name="motif-size"><label>7-24bp</label></div></div>
                            <div class="field"><div class="ui radio checkbox"><input type="radio" value="25to50bp_motifs" name="motif-size"><label>25-50bp</label></div></div>
                        </div>
                    </div>
                </li>
                <li>
                    <div class="ui form">
                        <div class="inline fields">
                            <label style="min-width:150px">Genotype:</label>
                            <div style="min-width: 70px">[<a class="prev-button" id="prev-genotype" href="#">←</a> <a class="next-button" id="next-genotype" href="#">→</a>]</div>
                            <div class="field"><div class="ui radio checkbox"><input type="radio" value="all_genotypes" name="genotype" checked="checked"><label>all</label></div></div>
                            <div class="field"><div class="ui radio checkbox"><input type="radio" value="HET" name="genotype"><label>heterozygous</label></div></div>
                            <div class="field"><div class="ui radio checkbox"><input type="radio" value="HOM" name="genotype"><label>homozygous</label></div></div>
                            <div class="field"><div class="ui radio checkbox"><input type="radio" value="MULTI" name="genotype"><label>multi-allelic</label></div></div>
                        </div>
                    </div>
                </li>
                <li>
                    <div class="ui form">
                        <div class="inline fields">
                            <label style="min-width:150px">Coverage:</label>
                            <div style="min-width: 70px">[<a class="prev-button" id="prev-coverage" href="#">←</a> <a class="next-button" id="next-coverage" href="#">→</a>]</div>
                            <div class="field"><div class="ui radio checkbox"><input type="radio" value="40x" name="coverage" checked="checked"><label  style="white-space:nowrap">40x</label></div></div>
                            <div class="field"><div class="ui radio checkbox"><input type="radio" value="30x" name="coverage"><label style="white-space:nowrap">30x</label></div></div>
                            <div class="field"><div class="ui radio checkbox"><input type="radio" value="20x" name="coverage"><label style="white-space:nowrap">20x</label></div></div>
                            <div class="field"><div class="ui radio checkbox"><input type="radio" value="10x" name="coverage"><label style="white-space:nowrap">10x</label></div></div>
                            <div class="field"><div class="ui radio checkbox"><input type="radio" value="exome" name="coverage"><label style="white-space:nowrap">exome (85x)</label></div></div>
                        </div>
                    </div>
                </li>
                <li>
                    <div class="ui form">
                        <div class="inline fields">
                            <label style="min-width:150px">Pure or Interrupted:</label>
                            <div style="min-width: 70px">[<a class="prev-button" id="prev-interruptions" href="#">←</a> <a class="next-button" id="next-interruptions" href="#">→</a>]</div>
                            <div class="field"><div class="ui radio checkbox"><input type="radio" value="both" name="interruptions" checked="checked"><label>both</label></div></div>
                            <div class="field"><div class="ui radio checkbox"><input type="radio" value="pure_repeats" name="interruptions"><label>only pure repeats</label></div></div>
                            <div class="field"><div class="ui radio checkbox"><input type="radio" value="with_interruptions" name="interruptions"><label>only interrupted repeats</label></div></div>
                        </div>
                    </div>
                </li>
                <li id="li-q-threshold">
                    <div class="ui form">
                        <div class="inline fields">
                            <label style="min-width:150px">Q Threshold:</label>
                            <div style="min-width: 70px">[<a class="prev-button" id="prev-q-threshold" href="#">←</a> <a class="next-button" id="next-q-threshold" href="#">→</a>]</div>
                            <div class="field"><div class="ui radio checkbox"><input type="radio" value="Q0" name="q-threshold" checked="checked"><label>none</label></div></div>
                            <div class="field"><div class="ui radio checkbox"><input type="radio" value="Q10" name="q-threshold"><label>≥ 0.1</label></div></div>
                            <div class="field"><div class="ui radio checkbox"><input type="radio" value="Q50" name="q-threshold"><label>≥ 0.5</label></div></div>
                            <div class="field"><div class="ui radio checkbox"><input type="radio" value="Q90" name="q-threshold"><label>≥ 0.9</label></div></div>
                        </div>
                    </div>
                </li>
                <li id="li-tool">
                    <div class="ui form">
                        <div class="inline fields">
                            <label style="min-width:150px">Tool:</label>
                            <div class="ui checkbox"><input type="checkbox" id="exclude-all-no-call-loci" name="exclude-all-no-call-loci"></div>
                            <label for="exclude-all-no-call-loci" style="font-weight: normal; cursor: pointer;">Only include loci called by all 3 tools</label>
                        </div>
                    </div>
                </li>
                <li id="li-tool2">
                    <div class="ui form">
                        <div class="inline fields">
                            <label style="min-width:150px">Tool:</label>
                            <div style="min-width: 70px">[<a class="prev-button" class="prev-button" id="prev-tool" href="#">←</a> <a class="next-button" id="next-tool" href="#">→</a>]</div>
                            <div class="field"><div class="ui radio checkbox"><input type="radio" value="ExpansionHunter" name="tool" checked="checked"><label>ExpansionHunter</label></div></div>
                            <div class="field"><div class="ui radio checkbox"><input type="radio" value="GangSTR" name="tool"><label>GangSTR</label></div></div>
                            <div class="field"><div class="ui radio checkbox"><input type="radio" value="HipSTR" name="tool"><label>HipSTR</label></div></div>
                            <div class="ui checkbox"><input type="checkbox" id="exclude-tool-no-call-loci" name="exclude-tool-no-call-loci"></div>
                            <label for="exclude-tool-no-call-loci" style="font-weight: normal; cursor: pointer;">Exclude "No Call" loci</label>
                        </div>
                    </div>
                </li>
            </ul>
        </div>
        <div id="accuracy-vs-q-legend" style="padding-left:50px; padding-top:30px; display:none; max-width: 850px">
            <div style="float:right; padding: 0px 25px 25px 10px">[<a href="#" id="hide-accuracy-vs-q-legend">hide this text</a>]</div>
            This plot shows how different minimum quality score thresholds affect post-filtering accuracy for each tool.
            The top panel shows the overall number (and fraction) of alleles (y-axis) that didn't pass a given minimum quality score threshold (x-axis).
            The bottom panel shows the accuracy (y-axis) of alleles that did pass the threshold. Here, accuracy is defined as
            <b>(true positive alleles + true negative alleles)/(total alleles)</b> and computed as follows: <br />
            <b>True positive alleles:</b> where the tool's call exactly equals the true allele size and has a quality score above the minimum threshold<br />
            <b>True negative alleles:</b> where the tool's call doesn't equal the true allele size and has a quality score below the minimum threshold<br />
            <b>Total alleles:</b> the total number of genotyped alleles, including those above and below the quality score threshold<br />
            <br />
            For GangSTR and HipSTR, the plot uses quality scores provided by the tools themselves. Since ExpansionHunter doesn't output a
            single quality score, the plot compares two different quality scores computed based on ExpansionHunter output:<br/>
            <b>Q from CIs = 1/exp(4 * (confidence interval size)/(called allele size))</b>  based on the approach used by EnsemblTR<br />
            <b>Q from reads = the fraction of reads that support the allele size vs. all allele sizes considered by ExpansionHunter at the given locus</b>. This is computed
                using the information provide in ExpansionHunter's json output for each locus.<br />

        </div>
        <div id="accuracy-by-allele-size-legend" style="padding-left:50px; padding-top:30px; display:none;">
            <div style="float:right; padding: 0px 25px 25px 10px">[<a href="#" id="hide-accuracy-by-allele-size-legend">hide this text</a>]</div>
            <b>x-axis:</b> represents the true allele size. Positive values are
            STR expansions, negative values are STR contractions.<br />
            <br />
            <b>Gray</b> represents loci where the tool didn't produce a genotype (i.e. due to errors or insufficient coverage).<br />
            <b>Red</b> is for loci that the tool called as homozygous reference and so entirely missed the variant. <br />
            <b>Brown</b> is when the tool incorrectly called an allele as being heterozygous reference.<br />
            <b>Teal</b> is where it called a contraction when the true allele is an expansion or vice versa.<br />
            <b>Blue</b> is for underestimated alleles (darker blue means underestimated by more repeats).<br />
            <b>Green</b> represents alleles where the tool exactly matched the true allele size. <br />
            <b>Orange</b> is for alleles that the tool overestimated (darker orange means a larger overestimate).<br />
        </div>
    </div>
    <div style="margin-top:0px">
        <img id="tool-comparison-plot-viewer" src="https://storage.googleapis.com/str-truth-set/hg38/figures/accuracy_vs_Q/tool_accuracy_vs_Q.2to6bp_motifs.all_genotypes.40x.svg" />
    </div>
</div>


<!-- script section -->
<script src="https://code.jquery.com/jquery-3.1.1.min.js"  integrity="sha256-hVVnYaiADRTO2PzUGmuLJr8BLUSjGIZsDYGmIJLv2b8=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/semantic-ui/2.4.1/semantic.min.js" integrity="sha512-dqw6X88iGgZlTsONxZK9ePmJEFrmHwpuMrsUChjAw1mRUhUITE5QU9pkcSox+ynfLhL15Sv2al5A0LVyDCmtUw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script>
<script>

    $('.ui.checkbox').checkbox();
    $('.prev-button').click((e) => {
        const linkId = e.target.id
        const name = linkId.replace(/^prev-/, '')
        const radioButtons = [...document.querySelectorAll(`input[name="${name}"]`)];
        const current = radioButtons.findIndex(btn => btn.checked);
        let idx = (current - 1) % radioButtons.length
        if (idx < 0) {
            idx = radioButtons.length - 1
        }
        radioButtons[idx].click();
        radioButtons[idx].checked = true;
    });

    $('.next-button').click((e) => {
        const linkId = e.target.id
        const name = linkId.replace(/^next-/, '')
        const radioButtons = [...document.querySelectorAll(`input[name="${name}"]`)];
        const current = radioButtons.findIndex(btn => btn.checked);
        const idx = (current + 1) % radioButtons.length
        radioButtons[idx].click();
        radioButtons[idx].checked = true;
    });

    let hideAccuracyVsQLegend = false
    let hideAccuracyByAlleleSizeLegend = false
    let update_tool_comparison_plot_viewer_image_url = () => {
        const plotType = $('input[name="plot-type"]:checked').val()
        const motif_size = "." + $('input[name="motif-size"]:checked').val()
        const genotype = "." + $('input[name="genotype"]:checked').val()
        const coverage = "." + $('input[name="coverage"]:checked').val()
        let interruptions = "." + $('input[name="interruptions"]:checked').val()
        let qThreshold = "." + $('input[name="q-threshold"]:checked').val()
        const tool = "." + $('input[name="tool"]:checked').val()
        const excludeAllNoCallLoci = $('input[name="exclude-all-no-call-loci"]:checked').val() ? ".exclude_no_call_loci" : ""
        const excludeToolNoCallLoci = $('input[name="exclude-tool-no-call-loci"]:checked').val() ? ".exclude_no_call_loci" : ""

        if (interruptions == ".both") {
            interruptions = ""
        }
        if (qThreshold == ".Q0") {
            qThreshold = ""
        }

        $("#li-no-call-loci").toggle(plotType == "AccuracyVsQ")
        $("#li-q-threshold").toggle(plotType == "AccuracyByAlleleSize")
        $("#li-tool").toggle(plotType == "AccuracyVsQ")
        $("#li-tool2").toggle(plotType == "AccuracyByAlleleSize")


        $("#accuracy-vs-q-legend").toggle(!hideAccuracyVsQLegend && (plotType == "AccuracyVsQ"))
        $("#accuracy-by-allele-size-legend").toggle(!hideAccuracyByAlleleSizeLegend && (plotType == "AccuracyByAlleleSize"))

        let imgUrl = "https://storage.googleapis.com/str-truth-set/hg38/figures"
        if (plotType == "AccuracyVsQ") {
            imgUrl += `/accuracy_vs_Q/tool_accuracy_vs_Q${motif_size}${genotype}${interruptions}${coverage}${excludeAllNoCallLoci}.svg`
            $("#tool-comparison-plot-viewer").attr({
                "src": imgUrl,
                "height": "800",
                "max-width": "100%",
            })
        } else if (plotType == "AccuracyByAlleleSize") {
            imgUrl += `/accuracy_by_allele_size/tool_accuracy_by_true_allele_size${motif_size}${genotype}${interruptions}${coverage}${excludeToolNoCallLoci}${qThreshold}${tool}.svg`
            $("#tool-comparison-plot-viewer").attr({
                "src": imgUrl,
                "style": "max-width:87%",
            })
        }

    }

    $("input[type='radio']").on("change", update_tool_comparison_plot_viewer_image_url);
    $("input[type='checkbox']").on("change", update_tool_comparison_plot_viewer_image_url);

    $("#hide-accuracy-vs-q-legend").click(() => {
        hideAccuracyVsQLegend = true
        $("#accuracy-vs-q-legend").hide()
    })
    $("#hide-accuracy-by-allele-size-legend").click(() => {
        hideAccuracyByAlleleSizeLegend = true
        $("#accuracy-by-allele-size-legend").hide()
    })

    //init the view
    update_tool_comparison_plot_viewer_image_url()
    
</script>
</body>
</html>
